cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java

```diff
@@ -88,10 +86,12 @@ public LevelDBTableFactory(CConfiguration cConf) {
     this.metadataTableName = cConf.get(Constants.MessagingSystem.METADATA_TABLE_NAME);
     this.messageTableName = cConf.get(Constants.MessagingSystem.MESSAGE_TABLE_NAME);
     this.payloadTableName = cConf.get(Constants.MessagingSystem.PAYLOAD_TABLE_NAME);
+
+    // Cache LevelDB tables with weak reference.
+    // We shouldn't expire cache by time as the returned instance can be held indefinitely and we cannot
+    // have more than one instance of LevelDB table that points to the same directory due to LevelDB uses FS lock.
     this.messageTableCache = CacheBuilder.newBuilder()
-      .expireAfterAccess(1, TimeUnit.HOURS)
-      .removalListener((RemovalListener<TopicMetadata, LevelDBMessageTable>)
-                         removalNotification -> Closeables.closeQuietly(removalNotification.getValue()))
+      .weakValues()
```

Comment:
with this, the tables never get closed? Unless the tables have a finalize() method.

Reply:
See my comment in the actual table class

Reply:
I guess it's ok because the TMS tables only change when topics are added/removed.